### DISCUSSION TITLE
flashing firmware for a new board  from  sdcard

### DISCUSSION THREAD

**Discussion by LiZheYuan-privaty** [micropython/micropython#16030] [STM32 / Pyboard]
I design  a new board  by stm32f4  .now  I want to flashing firmware from sdcard  on  the board .  It's not an update operation, it's a first-time installation。[mboot](https://github.com/micropython/micropython/tree/master/ports/stm32/mboot)
/README.md  provide method is update, how can I flashing the firmware for the first time

**Comment by sosi-deadeye** [comment]
AFIK the MCU requires a custom bootloader to do tasks like flashing itself. The program should run in RAM, so it should be possible to flash the controller. But in the first place, you need something on the MCU which is able to access the SDCard and reading the file system. Put a JTAG-Connector on your board. Some PCB-Manufacturers offers a service to flash microcontrollers with a firmware.

**Comment by andrewleech** [comment]
STM's don't support flash-from-new directly from sdcard. They do come with a built in rom bootloader that supports a few sources though, depending on the chip, tpyically USB DFU and UART https://community.st.com/t5/stm32-mcus/how-to-utilize-stm32-system-rom-bootloader-commands/ta-p/605286

Or as mentioned, jtag/swd header can always be used.

`mboot` can be flashed via one of these methods, after which it does have an optional `fsload` mode which can be used for in-place upgrades from a new firmware on filesystem, however I don't believe it support sdcard, only spiflash.

**Comment by LiZheYuan-privaty** [comment]
thanks , I asked my professor yesterday , he provide a new way. i aslo want to know how can i choose sdcard automatic in a firmware ,such as if i have a sdcard on the board , stm32  read  file from sdcard ,if i do not have a sdcard on the board,  stm32  will automatic skip the initialize of the sdcard.

**Reply by LiZheYuan-privaty** [reply]
i solver this problem by connect MICROPY_HW_SDCARD_DETECT_PIN  in tf card is CD pin ,micropython is quicker than c  a lot !!!!!!!!!!!!!!!!!!!!!!!!!!!!!
